File: UKIEPC/2021/hike.cpp



#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vi;


vi order;

vi graph[200005];
int seen[200005];

void rec(int node, int first){
    seen[node] = 1;
    if(first == 1){
        order.push_back(node);
    }

    for(auto child: graph[node]){

        if(!seen[child]) rec(child, 1-first);
    }

    if(!first){
        order.push_back(node);
    }
}



int main(){

    int n,m;
    int x,y;
    cin  >> n >> m;

    for(int i=0; i <= n; i++)graph[i] = {};
    fill(seen, seen+n+1,0);

    for(int i=0; i < m; i++){
        cin >> x >> y;
        graph[x].push_back(y);
        graph[y].push_back(x);
    }

    rec(1,1);

    for(auto c: order) cout << c << " ";


    return 0;
}

File: UKIEPC/2021/hike.py
import sys

sys.setrecursionlimit(100000000)
from collections import defaultdict
n,m = list(map(int, input().split()))

graph = [[] for i in range(n+1)]

seen = [0 for i in range(n+1)]


def rec(node,first=1):
    seen[node] =1
    if first:
        print(node, end=" ")

    for child in graph[node]:
        if not seen[child]:
            ### Traverse this

            rec(child,1-first)
    ### Otherwise add to the order when unwinding

    if not first:
        print(node, end=" ")









for i in range(m):
    x,y = list(map(int, input().split()))
    graph[x].append(y)
    graph[y].append(x)


rec(1)

